### PR TITLE
Exports: add PropertyHints as optional Parameters

### PR DESCRIPTION
**Pull request by mio991**
As the title says I added PropertyHints as optional Parameters.

If you prefere another design tell me.

### CODE REVIEW

**Comment by mio991** [comment]
I can not reproduce the CI failure on my end.

File: godot-macros/src/derive_godot_class.rs [hunk 1]
```diff
@@ -248,6 +265,18 @@ fn make_exports_impl(class_name: &Ident, fields: &Fields) -> TokenStream {
             let setter = proc_macro2::Literal::from_str(&exported_field.setter).unwrap();
             let field_type = exported_field.field.ty.clone();
 
+            let ExportHint {
+                hint_type,
+                description,
+            } = exported_field.hint.clone().unwrap_or(ExportHint {
+                hint_type: ident("PROPERTY_HINT_NONE"),
+                description: "".to_string(),
+            });
```
Comment (Bromeon) [inline godot-macros/src/derive_godot_class.rs:274]: We could probably add a 2nd constructor `ExportHint::none()`, then we wouldn't need to repeat the specifics in codegen.
```suggestion
            } = exported_field.hint.clone().unwrap_or(ExportHint::none());
```

File: itest/godot/ManualFfiTests.gd [hunk 2]
```diff
@@ -35,7 +35,17 @@ func test_export():
...
+	
+	var props = []
+	
+	for prop in obj.get_property_list():
+		props.push_back(prop["name"])
+		if prop["name"] == "texture_val":
+			assert_eq(prop["hint"], PropertyHint.PROPERTY_HINT_RESOURCE_TYPE)
+			assert_eq(prop["hint_string"], "Texture2D")
+	
+	assert_that(props.has("texture_val"), "'texture_val' was tested")
```
Comment (Bromeon) [inline itest/godot/ManualFfiTests.gd:47]:
```suggestion
    var found = null
	for prop in obj.get_property_list():
		if prop["name"] == "texture_val":
			found = prop
			break
	
	assert_that(found, "'texture_val' was tested")
	assert_eq(found["hint"], PropertyHint.PROPERTY_HINT_RESOURCE_TYPE)
	assert_eq(found["hint_string"], "Texture2D")
```

File: godot-macros/src/derive_godot_class.rs [hunk 3]
```diff
@@ -248,6 +265,18 @@ fn make_exports_impl(class_name: &Ident, fields: &Fields) -> TokenStream {
...
+
+            let description = description
+                .trim_matches(|c| c == '\\' || c == '"')
+                .to_string();
```
Comment (Bromeon) [inline godot-macros/src/derive_godot_class.rs:278]: Maybe add a short comment on what this is doing.
The `to_string()` is needed, or can this be used directly in `quote! { ... }`?

**Review by Bromeon — COMMENTED** [review]
Thanks a lot!

bors try

**Comment by Bromeon** [comment]
Regarding CI failure: `Texture` is not in the [minimal set of classes](https://github.com/godot-rust/gdextension/blob/master/godot-codegen/src/lib.rs#L256-L300) enabled for integration tests. You may compile with different features locally (try `./check.sh` instead of `cargo` directly).

Could you use one of the listed classes for the test instead?

File: itest/godot/ManualFfiTests.gd [hunk 4]
```diff
@@ -35,7 +35,16 @@ func test_export():
 	var node = Node.new()
 	obj.object_val = node
 	assert_eq(obj.object_val, node)
-
+	
+	var texture_val_meta = obj.get_property_list().filter(
+		func(el)->bool:
+			return el["name"] == "texture_val"
+			).front()
```
Comment (Bromeon) [inline itest/godot/ManualFfiTests.gd:42]:
```suggestion
	var texture_val_meta = obj.get_property_list().filter(
		func(e): return e["name"] == "texture_val"
	).front()
```

File: godot-macros/src/derive_godot_class.rs [hunk 5]
```diff
@@ -176,17 +176,43 @@ struct ExportedField {
...
         let setter = parser.handle_lit_required("setter")?;
 
+        let hint = if let Some(hint_type) = parser.handle_ident("hint")? {
+            Some(ExportHint {
+                hint_type,
+                description: parser.handle_lit_required("hint_desc")?,
+            })
+        } else {
+            None
+        };
```
Comment (Bromeon) [inline godot-macros/src/derive_godot_class.rs:209]: Maybe you could use `map()` here?
```suggestion
        let hint = parser.handle_ident("hint")?
            .map(|hint_type| ExportHint {
                hint_type,
                description: parser.handle_lit_required("hint_desc")?,
            }).transpose()?
```

Not 100% sure if this is directly applicable, but [`transpose()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose) is quite a cool function 🙂

**Review by Bromeon — COMMENTED** [review]
Looks mostly good!

When you're ready, could you squash the commits to 1?

File: itest/godot/ManualFfiTests.gd (hunk 4 above)
Comment (Bromeon) [reply itest/godot/ManualFfiTests.gd:42]: The suggestion was reformatting and stripping of the return value, why did you mark it as resolved?

**Review by Bromeon — COMMENTED** [review]
One of the suggestions you marked as resolved despite not applying it 😛 

But it's a minor thing, so I can fix it later myself.
Thanks!

bors r+

File: itest/godot/ManualFfiTests.gd (hunk 4 above)
Comment (mio991) [reply itest/godot/ManualFfiTests.gd:42]: I don't know, sry.

Comment (Bromeon) [reply itest/godot/ManualFfiTests.gd:42]: All good, no worries! I don't want to spend too much time on these little issues 🙂

**Pull request merged**